src/Renci.SshNet/Session.cs

```diff
@@ -669,6 +672,129 @@ public void Connect()
             }
         }
 
+#if FEATURE_TAP
+        /// <summary>
+        /// Asynchronously connects to the server.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous connect operation.</returns>
+        /// <exception cref="SocketException">Socket connection to the SSH server or proxy server could not be established, or an error occurred while resolving the hostname.</exception>
+        /// <exception cref="SshConnectionException">SSH session could not be established.</exception>
+        /// <exception cref="SshAuthenticationException">Authentication of SSH session failed.</exception>
+        /// <exception cref="ProxyException">Failed to establish proxy connection.</exception>
+        public async Task ConnectAsync(CancellationToken cancellationToken)
+        {
+            if (IsConnected)
+                return;
+
+            try
+            {
+                // ToDo: Migrate to SemaphoreSlim to enable async awaiting
+                // ToDo: SHould we really limit this globally here, or leave it to the app?
+                AuthenticationConnection.Wait();
+
+                if (IsConnected)
+                    return;
+
+                //lock (this)   // ToDo: Why do we need double locking at all?
```

Comment:
The current implementation of connection limiting does not play well in the async world and this would need to be completely reengineered if we decide to support this functionality in the async version. If you agree, then I'll remove this from ```ConnectAsync``` and document in the XmlDoc that the caller needs to take care of this. Maybe we should add a topic about this in the general documentation?

Reply:
I have changed the code and updated XmlDoc.
Can you please suggest other places where we should document this?